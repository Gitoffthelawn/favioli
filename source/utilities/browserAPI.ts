/**
 * API for more platform-agnostic access to browser extension apis.
 * Since browers ext API is kind of shifting sands, let's not do too much
 * work to try and make 100% stable. But we can use this for the more stable
 * APIs to avoid putting (chrome || browser) everywhere
 *
 * @reference https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions
 * @reference https://developer.chrome.com/docs/extensions/reference
 * @reference https://docs.microsoft.com/en-us/microsoft-edge/extensions-chromium/developer-guide/api-support
 *
 * @todo figure out how to properly import/declare/export global type from definitelyTyped.
 * Also looks like definitelyTyped is only for "chrome" global, so would need to map to
 * "browser" for approx. firefox types.
 */

interface BrowserExtensionAPI {
  // See @todo:
  // deno-lint-ignore no-explicit-any
  [name: string]: any;
}

declare global {
  const browser: BrowserExtensionAPI;
  const chrome: BrowserExtensionAPI;
}

const CHROME = 'CHROME';
const FIREFOX = 'FIREFOX';

export type BrowserName = 'CHROME' | 'FIREFOX';

const browserAPI = isBrowser(CHROME) ? chrome : browser; // Default to Chromium

export default browserAPI;
export const BROWSER: { [name: string]: BrowserName } = { CHROME, FIREFOX };
export const broserName = isBrowser(CHROME) ? CHROME : FIREFOX;

/**
 * What browser is this?
 * @param {string} toCheck to check
 */
export function isBrowser(toCheck: BrowserName): boolean | string {
  let currentBrowser = CHROME;
  try {
    // Use try block, since userAgent not guaranteed to exist.
    // If fail, assume Chromium
    // deno-lint-ignore no-explicit-any
    const userAgent: string = (navigator as any)?.userAgent || '';
    if (userAgent.indexOf('Firefox') > 0) {
      currentBrowser = FIREFOX;
    }
  } catch (_) {
    // Do nothing
  }

  if (!toCheck) currentBrowser;
  if (toCheck === CHROME && currentBrowser === CHROME) return true;
  if (toCheck === FIREFOX && currentBrowser === FIREFOX) return true;
  return false;
}
