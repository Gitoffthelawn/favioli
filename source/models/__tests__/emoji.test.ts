import { assert, assertEquals, assertRejects } from 'std/asserts';
import { describe, it } from 'std/bdd';
import { assertSpyCall, assertSpyCalls, stub } from 'std/mock';
import { assertSnapshot } from 'std/snapshot';

import browserAPI from 'browser';

import {
  areEqualEmojis,
  createEmoji,
  getEmoji,
  getEmojis,
  saveEmoji,
} from '../emoji.ts';

it('createEmoji', async (t) => {
  await assertSnapshot(t, createEmoji('poro', 'poro://poro-url'));
});

describe('getEmoji', () => {
  it('should get emoji from local Emoji DB', async () => {
    const emoji = await getEmoji('grinning squinting face');
    assertEquals(emoji?.emoji, 'ðŸ˜†');
  });

  it('should get emoji from storage.sync', async () => {
    const storageStub = stub(browserAPI.storage.sync, 'get', () => {
      return Promise.resolve({
        'Custom Emoji: poro': createEmoji('poro', 'poro://poro-url'),
      });
    });
    const emoji = await getEmoji('poro');
    assertEquals(emoji?.imageURL, 'poro://poro-url');
    assertEquals(emoji?.description, 'poro');
    assertSpyCalls(storageStub, 1);
    assertSpyCall(storageStub, 0, {
      args: [['Custom Emoji: poro']],
    });
    storageStub.restore();
  });
});

describe('saveEmoji', () => {
  it('should set emoji to storage.sync', async () => {
    const storageStub = stub(
      browserAPI.storage.sync,
      'set',
      () => Promise.resolve(),
    );
    const emoji = createEmoji('poro', 'poro://poro-url');
    await saveEmoji(emoji);
    assertSpyCalls(storageStub, 1);
    assertSpyCall(storageStub, 0, {
      args: [{ 'Custom Emoji: poro': emoji }],
      returned: Promise.resolve(),
    });
    storageStub.restore();
  });

  it('should error if emoji exists in local DB', async () => {
    const storageStub = stub(
      browserAPI.storage.sync,
      'set',
      () => Promise.resolve(),
    );
    const emoji = createEmoji('grinning squinting face', 'poro://poro-url');
    const errorMessage = 'This Emoji Already Exists!';
    await assertRejects(() => saveEmoji(emoji), Error, errorMessage);
    storageStub.restore();
  });
});

it('getEmojis', async () => {
  const storageStub = stub(browserAPI.storage.sync, 'get', () => {
    return Promise.resolve([
      createEmoji('poro', 'poro://poro-url'),
      createEmoji('nother', 'nother://nother-custom-emoji'),
    ]);
  });
  const emojis = await getEmojis(['nother', 'poro', 'grinning squinting face']);
  assertEquals(emojis['grinning squinting face'].emoji, 'ðŸ˜†');
  assertEquals(emojis['poro'].imageURL, 'poro://poro-url');
  assertEquals(emojis['nother'].imageURL, 'nother://nother-custom-emoji');

  assertSpyCalls(storageStub, 1);

  const expectedArgs = [['Custom Emoji: nother', 'Custom Emoji: poro']];
  assertSpyCall(storageStub, 0, { args: expectedArgs });
  storageStub.restore();
});

describe('areEqualEmojis', async () => {
  const emoji_1a = await getEmoji('grinning squinting face');
  const emoji_1b = await getEmoji('grinning squinting face');
  const emoji_2 = await getEmoji('tractor');
  const customEmoji_1a = createEmoji('poro', 'poro://poro-url');
  const customEmoji_1b = createEmoji('poro', 'poro://poro-url');

  if (!emoji_1a || !emoji_1b || !emoji_2) throw new Error('Bad emoji creation');

  it('emoji_1a === emoji_1b', () => assert(areEqualEmojis(emoji_1a, emoji_1b)));
  it('emoji_1a !== emoji_2', () => assert(!areEqualEmojis(emoji_1a, emoji_2)));
  it('emoji_1b !== emoji_2', () => assert(!areEqualEmojis(emoji_1b, emoji_2)));
  it('customEmoji_1a === customEmoji_1b', () =>
    assert(areEqualEmojis(customEmoji_1a, customEmoji_1b)));
  it('customEmoji_1a !== emoji_1a', () =>
    assert(!areEqualEmojis(customEmoji_1a, emoji_1a)));
});
