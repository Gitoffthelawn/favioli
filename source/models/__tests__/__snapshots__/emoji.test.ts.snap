export const snapshot = {};

snapshot[`createEmoji 1`] = `
{
  aliases: [
    "poro",
  ],
  description: "poro",
  emoji: "",
  emojiVersion: 0,
  group: "Custom Emojis",
  imageURL: "poro://poro-url",
  subgroup: "custom-emojis",
  tags: [
  ],
  unicodeVersion: 0,
}
`;
